[Gary Yansen] iP

g4ryy opened the pull request:
# Nyx Chatbot :robot:

>“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Nyx isn't an ordinary chatbot. It helps you to manage your tasks and events, and most importantly it's
*  Easy to use  
* Responsive 
* ~~FAST~~ **SUPER FAST** to use :zap:

With Nyx, you no longer have to worry about forgetting important tasks or events! 

Nyx supports the following functionalities:

- [x] Add up to 3 different types of tasks: _Todo_, _Deadline_, _Event_
- [x] Mark task as done
- [x] Delete task
- [x] List all tasks
- [x] Search for tasks using keyword 

### Getting Started

1. Download the latest version from [here](https://github.com/g4ryy/ip/releases/tag/v0.1)
2. Follow the instructions [here](https://github.com/g4ryy/ip)
3. Start chatting with Nyx :zany_face:

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
``` Java
public static void main(String[] args) {
    new Nyx("data", "nyx.txt").run();
}
```

Commit:
Add Gradle support

g4ryy committed:
Complete level 1

g4ryy committed:
Add Task.java

g4ryy committed:
Complete level 2

g4ryy committed:
Complete level 3

g4ryy committed:
Add Todo.java

g4ryy committed:
Add Deadline.java

g4ryy committed:
Add Event.java

g4ryy committed:
Complete level 4

g4ryy committed:
Complete automated text UI testing

g4ryy committed:
Complete level 5

g4ryy committed:
Add new input command and update expected output

g4ryy committed:
Complete level 6

g4ryy committed:
Task class: change to abstract class

g4ryy committed:
Create respond method to handle output logic

g4ryy committed:
Add DataManager to handle save and load of data

g4ryy committed:
Overload constructor and add dataFormat method in Task and its subclasses

g4ryy committed:
NyxException class: change parent class to Exception

g4ryy committed:
Complete level 7

g4ryy committed:
Store deadline and event dates as a java.time.LocalDate object

g4ryy committed:
Merge branch 'branch-Level-7' into master

g4ryy committed:
Merge branch 'branch-Level-8' into master

g4ryy committed:
Update input.txt and EXPECTED.txt

g4ryy committed:
Refactor code to extract out closely related code as classes

g4ryy committed:
Organize the classes into nyx package

g4ryy committed:
Add JUnit Tests for Task, Deadline, Event, TaskList and ToDo

g4ryy committed:
Package the app as a JAR file

g4ryy committed:
Add JavaDoc comments

g4ryy committed:
Merge branch 'branch-A-JavaDoc' into master

g4ryy committed:
Tweak the code to comply with coding standards

g4ryy committed:
Parser class: remove curly braces for case statements

g4ryy committed:
Add search functionality

g4ryy committed:
Merge branch 'branch-A-CodingStandard' into master

g4ryy committed:
Merge branch 'branch-Level-9' into master

ryantianj commented on src/main/java/Nyx.java:
```diff
@@ -0,0 +1,65 @@
+    private final Ui ui;
+    private TaskList taskList;
+    private final Storage storage;
+
```
Great use of final modifiers! Makes it really clear that these Objects will not be changed.

ryantianj commented on src/main/java/nyx/Task.java:
```diff
@@ -0,0 +1,54 @@
+     * Returns a String representation of the task in the format required for saving into hard disk.
+     * @return String representation of the task in the format required to save into text hard disk.
+     */
+    public abstract String dataFormat();
```
Perhaps naming it with a verb, such as formatData,  would be better.

rohit0718 replied:
perhaps `formattedData` might be more appropriate here since it returns a string. what do you think @ryantianj?

ryantianj commented on src/main/java/nyx/TaskList.java:
```diff
@@ -0,0 +1,113 @@
+     * in the format required for saving into hard disk.
+     * @return String representation of all the tasks in the list in the format required for saving into hard disk.
+     */
+    public String genSaveFormat() {
```
Personally, 'gen' was not very clear to me (generate?), may be better to write in full (generateSaveFormat)

rohit0718 replied:
I agree. Perhaps can even use something simpler like `getSaveFormat`

ryantianj commented on src/main/java/nyx/TaskList.java:
```diff
@@ -0,0 +1,113 @@
+    @Override
+    public String toString() {
+        if (tasks.isEmpty()) {
+            return "You do not have any task currently";
```
Minor grammar, perhaps it should be "You do not have any task**s** currently"

ryantianj commented on src/main/java/nyx/Task.java:
```diff
@@ -0,0 +1,54 @@
+        return isDone ? 1 : 0;
+    }
+
+    String getContent() {
```
Perhaps adding an access modifier here would make this method easier to read!

rohit0718 replied:
i think the author intended it to be package private here, but i might be mistaken. @g4ryy is this method being used elsewhere in your package?

g4ryy replied:
Hi @rohit0718, thanks for pointing this out! Indeed, my intention is to make this method package private since it is not used outside the package.

ryantianj reviewed (commented):
Honestly, your code formatting and quality was excellent. It was very difficult to find faults within your code and I learnt quite a bit (like OOP design) reading your code. I had to really nitpick your code to find faults with it. Hope it will be useful for you!

rohit0718 commented on src/main/java/Nyx.java:
```diff
@@ -0,0 +1,65 @@
+            + "|_|    \\__| /_/    /_/   \\_\\\n";
+    private final Ui ui;
+    private TaskList taskList;
+    private final Storage storage;
```
super small thing, but perhaps can grp the variables with the similar prefix together! in this case that would be:
```suggestion
    private final Ui ui;
    private final Storage storage;
    private TaskList taskList;
```

rohit0718 commented on src/main/java/Nyx.java:
```diff
@@ -0,0 +1,65 @@
+        ui.displayStart(logo);
+
+        String input;
+
```
perhaps can remove the spacing here since its not needed.

rohit0718 commented on src/main/java/Nyx.java:
```diff
@@ -0,0 +1,65 @@
+    public static void main(String[] args) {
+        new Nyx("data", "nyx.txt").run();
+    }
+}
```
perhaps you can standardize all files to include a new line!

rohit0718 commented on src/main/java/nyx/Deadline.java:
```diff
@@ -0,0 +1,53 @@
+ */
+public class Deadline extends Task {
+    private static final String DATETIME_FORMAT = "yyyy-MM-dd H:m";
+    private final LocalDateTime by;
```
perhaps a better name for the variable might be `endDateTime`. it is not immediately obvious what `by` refers to!

rohit0718 commented on src/main/java/nyx/Event.java:
```diff
@@ -0,0 +1,53 @@
+ */
+public class Event extends Task {
+    private static final String DATETIME_FORMAT = "yyyy-MM-dd H:m";
+    private final LocalDateTime at;
```
same for this, perhaps can be `startDateTime`!

rohit0718 commented on src/main/java/nyx/Storage.java:
```diff
@@ -0,0 +1,96 @@
+        String[] splitLine = line.split(", ");
+
+        boolean isDone = splitLine[1].equals("1");
+
+        Task task;
+
```
same comment as above, perhaps can remove the spacing here.

rohit0718 reviewed (commented):
Overall coding style was great! Only have some minor comments!

g4ryy committed:
Minor changes to improve code quality

justintanyf commented on src/main/java/nyx/Deadline.java:
```diff
@@ -0,0 +1,53 @@
+    private final LocalDateTime by;
+
+    /**
+     * Constructs a deadline task with its description, deadline and an indicator of whether it is marked as done.
```
The javaDocs convention said that we should leave a gap between the description and @param

justintanyf reviewed (commented):
Overall an excellent job!

g4ryy committed:
Merge remote-tracking branch 'origin/add-gradle-support' into master

g4ryy committed:
Set up gradle to automate project builds

g4ryy committed:
Merge branch 'branch-A-Gradle' into master

g4ryy committed:
Add configuration files for Checkstyle

g4ryy committed:
Fix coding style violations

g4ryy committed:
Fix coding style violations in test code

g4ryy committed:
Merge branch 'branch-A-CheckStyle' into master

g4ryy committed:
Implement GUI for Nyx

g4ryy committed:
Merge branch 'branch-Level-10' into master

g4ryy committed:
Make dialog box display full content

Users are unable to view all the tasks as the dialog box get cut when there are
many tasks.

Let's add a minHeight attribute with value = "-Infinity" to the Label
component in Dialogbox.fxml.

g4ryy committed:
Fix output time format

g4ryy committed:
Change GUI appearances

g4ryy committed:
Add assertions to document important assumptions

g4ryy committed:
Organize Task, ToDo, Deadline, Event and TaskList classes into nyx.task package

g4ryy committed:
Move ui package into nyx package

g4ryy committed:
Extract *Command classes that inherits from abstract Command class

Parser's parse method handles all the logic of interpreting user input and executing all commands,
which makes the method very long.

Breaking up the logic into a few classes and methods enhances
readability and makes it easier to debug.

Let's,
* create an abstract Command class
* create several classes that inherits from the Command class and
deals with operations specific to that command.

g4ryy committed:
Merge pull request #2 from g4ryy/branch-A-CodeQuality

Refactor code to improve code quality

g4ryy committed:
Merge branch 'master' of https://github.com/g4ryy/ip into master

g4ryy committed:
Merge branch 'master' into branch-A-Assertions

g4ryy committed:
Merge pull request #1 from g4ryy/branch-A-Assertions

Add assertions

g4ryy committed:
Update code to handle incorrect deadline and event formats

g4ryy committed:
Make command keywords case insensitive

Command keywords are case sensitive.

A case insensitive keyword is more user-friendly because users cannot be
expected to remember the exact case of the keywords.

Let's update the parse method in Parser class to use case-insensitive keywords

g4ryy committed:
Add update command to modify existing tasks

g4ryy committed:
Merge branch 'branch-feature-update' into master

g4ryy committed:
Update GUI

g4ryy committed:
Set theme jekyll-theme-hacker

g4ryy committed:
Add user guide and product screenshot

g4ryy committed:
Merge branch 'master' of https://github.com/g4ryy/ip into master

g4ryy committed:
Fix checkstyle error

g4ryy committed:
Improve code quality

g4ryy committed:
Set theme jekyll-theme-time-machine

g4ryy committed:
Set theme jekyll-theme-cayman

g4ryy committed:
Update README.md

g4ryy committed:
Fix Javadoc comments

g4ryy committed:
Merge branch 'master' of https://github.com/g4ryy/ip into master

g4ryy committed:
Update README.md

g4ryy committed:
Update README.md
